src/ray/common/ray_syncer/ray_syncer.cc

```diff
@@ -350,14 +354,31 @@ ServerBidiReactor *RaySyncerService::StartSync(grpc::CallbackServerContext *cont
       syncer_.GetIOContext(),
       syncer_.GetLocalNodeID(),
       [this](auto msg) mutable { syncer_.BroadcastMessage(msg); },
-      [this](const std::string &node_id, bool reconnect) mutable {
+      [this](RaySyncerBidiReactor *reactor, bool reconnect) mutable {
         // No need to reconnect for server side.
         RAY_CHECK(!reconnect);
+        const auto &node_id = reactor->GetRemoteNodeID();
+        if (syncer_.sync_reactors_.contains(node_id) &&
+            syncer_.sync_reactors_.at(node_id) != reactor) {
+          // There is a new connection to the node, no need to clean up.
+          // This can happen when there is transient network error and the client
+          // reconnects. The sequence of events are:
+          // 1. Client reconnects, StartSync is called
+          // 2. syncer_.Disconnect is called and the old reactor is removed from
+          // sync_reactors_
+          // 3. syncer_.Connect is called and the new reactor is added to sync_reactors_
+          // 4. OnDone method of the old reactor is called which calls this cleanup_cb_
+          return;
+        }
         syncer_.sync_reactors_.erase(node_id);
         syncer_.node_state_->RemoveNode(node_id);
       });
   RAY_LOG(INFO).WithField(kLogKeyNodeID, NodeID::FromBinary(reactor->GetRemoteNodeID()))
       << "Get connection";
+  // Disconnect exiting connection if there is any.
+  // This can happen when there is transient network error
+  // and the client reconnects.
+  syncer_.Disconnect(reactor->GetRemoteNodeID());
```

Comment:
QQ: why don't we call this in clean_cb?

Reply:
`syncer_.Disconnect` triggers the call of `cleanup_cb`